src/solvers/dgsem_structured/dg_3d.jl

```diff
@@ -263,6 +263,84 @@ end
+# [arXiv: 2008.12044v2] (https://arxiv.org/pdf/2008.12044)
+@inline function calcflux_fv!(fstar1, fstar2, fstar3, u::AbstractArray{<:Any,5},
+                              mesh::Union{StructuredMesh{3}, P4estMesh{3}}, nonconservative_terms::Val{false},
+                              equations, volume_flux_fv, dg::DGSEM, element, cache)
+  @unpack contravariant_vectors = cache.elements
+  @unpack weights, derivative_matrix = dg.basis
+
+  fstar1[:, 1,            :, :] .= zero(eltype(fstar1))
+  fstar1[:, nnodes(dg)+1, :, :] .= zero(eltype(fstar1))
+
+  for k in eachnode(dg), j in eachnode(dg)
+    normal_direction = get_contravariant_vector(1, contravariant_vectors, 1, j, k, element)
+
+    for i in 2:nnodes(dg)
+      u_ll = get_node_vars(u, equations, dg, i-1, j, k, element)
+      u_rr = get_node_vars(u, equations, dg, i,   j, k, element)
+
+      for ii in 1:nnodes(dg)
+        normal_direction += weights[i-1] * derivative_matrix[i-1, ii] * get_contravariant_vector(1, contravariant_vectors, ii, j, k, element)
+      end
+
+      # Compute the contravariant flux
+      contravariant_flux1 = volume_flux_fv(u_ll, u_rr, normal_direction, equations)
+
+      set_node_vars!(fstar1, contravariant_flux1, equations, dg, i, j, k)
+    end
+  end
+
+  fstar2[:, :, 1           , :] .= zero(eltype(fstar2))
+  fstar2[:, :, nnodes(dg)+1, :] .= zero(eltype(fstar2))
+
+  for k in eachnode(dg), i in eachnode(dg)
+    normal_direction = get_contravariant_vector(2, contravariant_vectors, i, 1, k, element)
+
+    for j in 2:nnodes(dg)
+      u_ll = get_node_vars(u, equations, dg, i, j-1, k, element)
+      u_rr = get_node_vars(u, equations, dg, i, j,   k, element)
+
+      for ii in 1:nnodes(dg)
+        normal_direction += weights[j-1] * derivative_matrix[j-1, ii] * get_contravariant_vector(2, contravariant_vectors, i, ii, k, element)
+      end
+
+      # Compute the contravariant flux
+      contravariant_flux1 = volume_flux_fv(u_ll, u_rr, normal_direction, equations)
+
+      set_node_vars!(fstar2, contravariant_flux1, equations, dg, i, j, k)
+    end
+  end
+
+  fstar3[:, :, :, 1           ] .= zero(eltype(fstar3))
+  fstar3[:, :, :, nnodes(dg)+1] .= zero(eltype(fstar3))
+
+  for j in eachnode(dg), i in eachnode(dg)
+    normal_direction = get_contravariant_vector(3, contravariant_vectors, i, j, 1, element)
+
+    for k in 2:nnodes(dg)
+      u_ll = get_node_vars(u, equations, dg, i, j, k-1, element)
+      u_rr = get_node_vars(u, equations, dg, i, j, k,   element)
+
+      for ii in 1:nnodes(dg)
+        normal_direction += weights[k-1] * derivative_matrix[k-1, ii] * get_contravariant_vector(3, contravariant_vectors, i, j, ii, element)
+      end
+
+      # Compute the contravariant flux
+      contravariant_flux1 = volume_flux_fv(u_ll, u_rr, normal_direction, equations)
+
+      set_node_vars!(fstar3, contravariant_flux1, equations, dg, i, j, k)
+    end
+  end
```

Comment:
It would be great if one of the original authors of the preprint could have a look

Reply:
Could you please choose someone and ping them, @Cczernik?